From 4353e31fee92683dd436f1564423f32223a6d68c Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Thu, 09 Sep 2021 20:06:30 +0300
Subject: [PATCH] [SW-55646] gaudi2: arm: allow asic fit in spsram

Fit image can reside not only in SRAM but also in SPSRAM (used
with ELBI).
As to pldm and onix, it only converts from arm view to absolute.
This patch differentiate between the configuration by moving
the implementation from common to arch specific.

Change-Id: I80219b4e2e0c6499af1142748b315298e3fe4e24
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/arch/gaudi2/arm/asic.c b/arch/gaudi2/arm/asic.c
index a3f6d20..4d3b111 100644
--- a/arch/gaudi2/arm/asic.c
+++ b/arch/gaudi2/arm/asic.c
@@ -13,6 +13,17 @@
 #include <pci.h>
 #include <address_space.h>
 
+#define SDRAM_BASE_FULL 0x1000FFFFFD000000ull
+
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	/* Return 64 bit full address */
+	if ((base >= HL_SP_SRAM_BASE) && (base <= HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE))
+		return ((base & ~HL_SP_SRAM_BASE) | HL_SP_SRAM_BASE_FULL);
+
+	return ((base & ~SDRAM_BASE) | SDRAM_BASE_FULL);
+}
+
 u32 hl_skip_pcie_init(void)
 {
 	/* TODO: Use DumpSel[*] to skip PCIe initialization */
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index f9bf398..bed2d38 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -50,12 +50,6 @@
 
 #define HL_PCIE_LINK_RETRIES		50
 
-u64 lkd_dev_spec_get_img_addr(uintptr_t base)
-{
-	/* Return 64 bit full address */
-	return ((base ^ SDRAM_BASE) | SDRAM_BASE_FULL);
-}
-
 #define mmPCIE_WRAP_VUART_TX_DATA	0x4C14F60
 #define mmPCIE_WRAP_VUART_TX_STATUS	0x4C14F64
 
diff --git a/arch/gaudi2/arm/onix.c b/arch/gaudi2/arm/onix.c
index a0ef4b4..b0eee83 100644
--- a/arch/gaudi2/arm/onix.c
+++ b/arch/gaudi2/arm/onix.c
@@ -10,6 +10,13 @@
 
 #define SYS_CLOCK_FREQUENCY	35000000ULL
 
+#define SDRAM_BASE_FULL	0x1001000200000000ull
+
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	return ((base & ~SDRAM_BASE) | SDRAM_BASE_FULL);
+}
+
 u32 hl_spi_baudr_get(void)
 {
 	u32 factor;
diff --git a/arch/gaudi2/arm/pldm.c b/arch/gaudi2/arm/pldm.c
index 0548343..41e6d82 100644
--- a/arch/gaudi2/arm/pldm.c
+++ b/arch/gaudi2/arm/pldm.c
@@ -8,6 +8,12 @@
 
 #include <preboot.h>
 
+#define SDRAM_BASE_FULL 0x1000FFFFFD000000ull
+
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	return ((base & ~SDRAM_BASE) | SDRAM_BASE_FULL);
+}
 
 u32 hl_skip_pcie_init(void)
 {
